src/workerd/jsg/jsg.c++

```diff
@@ -278,27 +278,68 @@ JsSymbol Lock::symbolAsyncDispose() {
   return IsolateBase::from(v8Isolate).getSymbolAsyncDispose();
 }
 
-void Lock::adjustExternalMemory(ssize_t amount) {
-  v8Isolate->AdjustAmountOfExternalAllocatedMemory(static_cast<int64_t>(amount));
+void ExternalMemoryAdjuster::maybeDeferAdjustment(v8::Isolate* isolate, size_t amount) {
+  if (isolate == nullptr) return;
+  if (v8::Locker::IsLocked(isolate)) {
+    isolate->AdjustAmountOfExternalAllocatedMemory(static_cast<int64_t>(-amount));
+  } else {
+    // Otherwise, if we don't have the isolate locked, defer the adjustment to the next
+    // time that we do.
+    auto& jsgIsolate = *reinterpret_cast<IsolateBase*>(isolate->GetData(0));
+    jsgIsolate.deferExternalMemoryDecrement(static_cast<int64_t>(amount));
+  }
+}
+
+ExternalMemoryAdjuster::ExternalMemoryAdjuster(v8::Isolate* isolate, size_t amount)
+    : amount(amount), isolate(isolate) {
+  KJ_DASSERT(isolate != nullptr);
+  isolate->AdjustAmountOfExternalAllocatedMemory(amount);
 }
 
-namespace {
-struct ExternalMemoryAdjuster {
-  int64_t size;
-  v8::Isolate* isolate;
-  ExternalMemoryAdjuster(v8::Isolate* isolate, size_t size)
-      : size(static_cast<int64_t>(size)), isolate(isolate) {
-    isolate->AdjustAmountOfExternalAllocatedMemory(size);
+ExternalMemoryAdjuster::ExternalMemoryAdjuster(ExternalMemoryAdjuster&& other) :
+    amount(other.amount), isolate(other.isolate) {
+  other.amount = 0;
+  other.isolate = nullptr;
+}
+
+ExternalMemoryAdjuster& ExternalMemoryAdjuster::operator=(ExternalMemoryAdjuster&& other) {
+  // If we currently have an amount, adjust it back to zero.
+  if (amount > 0) isolate->AdjustAmountOfExternalAllocatedMemory(-static_cast<int64_t>(amount));
+  amount = other.amount;
+  isolate = other.isolate;
+  other.amount = 0;
+  other.isolate = nullptr;
+  return *this;
+}
+
+ExternalMemoryAdjuster::~ExternalMemoryAdjuster() {
+  if (amount != 0) {
+    maybeDeferAdjustment(isolate, amount);
   }
+}
 
-  ~ExternalMemoryAdjuster() noexcept(false) {
-    isolate->AdjustAmountOfExternalAllocatedMemory(-size);
+void ExternalMemoryAdjuster::adjust(Lock& js, ssize_t amount) {
+  KJ_ASSERT(isolate == js.v8Isolate, "wrong isolate?");
+  int64_t adjustment = 0;
+  if (amount > 0) {
+    this->amount += amount;
+    adjustment = static_cast<int64_t>(amount);
+  } else if (amount < 0) {
+    size_t previousAmount = this->amount;
+    size_t amountToDecrement = kj::min(this->amount, -amount);
+    this->amount -= amountToDecrement;
+    KJ_ASSERT(this->amount < previousAmount);  // Be overly cautious about arithmetic overflow.
+    adjustment = -static_cast<int64_t>(amountToDecrement);
```

Comment:
My code above throws an exception if an adjustment would make the adjuster's size negative, since that's probably a bug. But if you really want to clamp the size to zero, then it could be...

```cpp
  amount = kj::max(amount, -this->amount);
  this->amount += amount;
  isolate->AdjustAmountOfExternalAllocatedMemory(amount);
```